File: lib/cis/suppress.rb
# frozen_string_literal: true

require "yaml"

module Cis
  # Operator-declared exclusions, read from config/suppress.yml.
  #
  #   suppress:
  #     - cloud:    "aws"            # cloud name, or "*" for every cloud
  #       control:  "6.3"            # control id, or a glob like "6.*"
  #       resource: "sg-0a1b2c3d4e5f60708"   # optional; matched against the
  #                                         # finding's evidence text
  #       reason:   "known exception - legacy jump host"
  #
  # Suppressed findings are reported with status SUPPRESSED and do not trip
  # the scan exit-code gate.
  class Suppressions
    def self.load(path = File.join(Cis::ROOT, "config", "suppress.yml"))
      rules = []
      if File.exist?(path)
        data = YAML.safe_load(File.read(path), aliases: true) || {}
        rules = Array(data["suppress"])
      end
      new(rules)
    end

    attr_reader :rules

    def initialize(rules)
      @rules = rules
    end

    def apply(findings, cloud)
      findings.map do |f|
        next f unless match?(f, cloud)

        reason = reason_for(f, cloud)
        f.merge(
          "status"     => "SUPPRESSED",
          "suppressed" => true,
          "evidence"   => "#{f['evidence']} [suppressed: #{reason}]"
        )
      end
    end

    def match?(finding, cloud)
      @rules.any? { |r| rule_hits?(r, finding, cloud) }
    end

    def rule_hits?(rule, finding, cloud)
      return false unless rule["cloud"] == "*" || rule["cloud"] == cloud
      return false unless glob_match(rule["control"], finding["id"].to_s)

      res = rule["resource"]
      res.nil? || res.empty? ||
        finding["evidence"].to_s.include?(res) || finding["id"].to_s.include?(res)
    end

    def reason_for(finding, cloud)
      rule = @rules.find { |r| rule_hits?(r, finding, cloud) }
      rule && rule["reason"] ? rule["reason"] : "suppressed in config/suppress.yml"
    end

    def glob_match(pattern, value)
      return true if pattern.nil? || pattern == "*"
      return false unless pattern.is_a?(String)

      re = Regexp.new("\\A" + pattern.gsub(".", "\\.").gsub("*", ".*") + "\\z")
      re.match?(value)
    end
  end
end

File: cis_cloud/suppress.py
"""Operator-declared exclusions, read from config/suppress.yml."""

from __future__ import annotations

import re
from pathlib import Path
from typing import Optional

import yaml

from . import get_root


class Suppressions:
    def __init__(self, rules: list):
        self.rules = rules

    @classmethod
    def load(cls, path: Optional[Path] = None) -> "Suppressions":
        path = path or (get_root() / "config" / "suppress.yml")
        rules = []
        if path.exists():
            data = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
            rules = list(data.get("suppress") or [])
        return cls(rules)

    def apply(self, findings: list[dict], cloud: str) -> list[dict]:
        out = []
        for f in findings:
            if not self.match(f, cloud):
                out.append(f)
                continue
            reason = self.reason_for(f, cloud)
            out.append({
                **f,
                "status": "SUPPRESSED",
                "suppressed": True,
                "evidence": f"{f.get('evidence', '')} [suppressed: {reason}]",
            })
        return out

    def match(self, finding: dict, cloud: str) -> bool:
        return any(self._rule_hits(r, finding, cloud) for r in self.rules)

    def _rule_hits(self, rule: dict, finding: dict, cloud: str) -> bool:
        if rule.get("cloud") != "*" and rule.get("cloud") != cloud:
            return False
        if not self._glob_match(rule.get("control"), str(finding.get("id", ""))):
            return False
        res = rule.get("resource")
        if res is None or res == "":
            return True
        # Prefer the structured `resource` field when present; fall back to
        # matching against the evidence text.
        return (res in str(finding.get("resource", "") or "")
                or res in str(finding.get("evidence", ""))
                or res in str(finding.get("id", "")))

    def reason_for(self, finding: dict, cloud: str) -> str:
        for r in self.rules:
            if self._rule_hits(r, finding, cloud):
                return r.get("reason") or "suppressed in config/suppress.yml"
        return "suppressed in config/suppress.yml"

    @staticmethod
    def _glob_match(pattern, value: str) -> bool:
        if pattern is None or pattern == "*":
            return True
        if not isinstance(pattern, str):
            return False
        re_src = r"^" + pattern.replace(".", r"\.").replace("*", ".*") + r"$"
        return re.match(re_src, value) is not None
